### PR TITLE
Harmonized error/logProbability in entire Factor/Conditional hierarchy

### PR DESCRIPTION
**Pull request by dellaert**
Cleaning up before releasing 4.2 and implementing last hybrid PR
- renamed logDensity to logProbability, as density is continuous only
- we now have logProbability(x) = log(evaluate(x)) = K - error(x) for all conditionals
- probPrime(x) = exp(-error(x)) as always, but is only supposed to be used in factors

My advice for reviewing (seems daunting but most is rote):
- start with Factor.h, Conditional.h and the rest of inference - read the updated docs (and please flag issues).
- Then hybrid.
- Then linear, discrete, symbolic

Note that some methods are gone as now implemented in bottom base classes.

### CODE REVIEW

File: gtsam/inference/Factor.h [hunk 1]
```diff
@@ -128,6 +145,12 @@ typedef FastSet<FactorIndex> FactorIndexSet;
    /** Iterator at end of involved variable keys */
    const_iterator end() const { return keys_.end(); }
 
+  /**
+   * All factor types need to implement an error function.
+   * In factor graphs, this is the negative log-likelihood.
+   */
+  virtual double error(const HybridValues& c) const;
```
Comment (ProfFan) [inline gtsam/inference/Factor.h:152]: indent

File: gtsam/symbolic/SymbolicConditional.h [hunk 2]
```diff
@@ -17,10 +17,10 @@
 
 #pragma once
 
-#include <gtsam/symbolic/SymbolicFactor.h>
-#include <gtsam/inference/Conditional.h>
 #include <gtsam/base/Testable.h>
 #include <gtsam/base/types.h>
+#include <gtsam/inference/Conditional-inst.h>
```
Comment (ProfFan) [inline gtsam/symbolic/SymbolicConditional.h:22]: `inst` headers should not be included?

**Review by ProfFan — APPROVED** [review]
Most looks great to me!

File: gtsam/hybrid/HybridBayesNet.cpp [hunk 3]
```diff
@@ -318,33 +291,28 @@ HybridValues HybridBayesNet::sample() const {
...
-double HybridBayesNet::error(const HybridValues &values) const {
-  GaussianBayesNet gbn = choose(values.discrete());
-  return gbn.error(values.continuous());
-}
-
-/* ************************************************************************* */
-AlgebraicDecisionTree<Key> HybridBayesNet::error(
+AlgebraicDecisionTree<Key> HybridBayesNet::logProbability(
     const VectorValues &continuousValues) const {
   AlgebraicDecisionTree<Key> error_tree(0.0);
```
Comment (varunagrawal) [inline gtsam/hybrid/HybridBayesNet.cpp:296]: Rename `error_tree` to `log_prob_tree`?

File: gtsam/hybrid/HybridValues.h [hunk 4]
```diff
@@ -37,12 +37,15 @@ namespace gtsam {
...
  private:
-  // VectorValue stored the continuous components of the HybridValues.
+  /// Continuous multi-dimensional vectors for \class GaussianFactor.
   VectorValues continuous_;
 
-  // DiscreteValue stored the discrete components of the HybridValues.
+  /// Discrete values for \class DiscreteFactor.
   DiscreteValues discrete_;
 
+  /// Continuous, differentiable manifold values for \class NonlinearFactor.
```
Comment (varunagrawal) [inline gtsam/hybrid/HybridValues.h:46]: We should have `HybridVectorValues` and `HybridValues`. This is very unclean.

File: gtsam/inference/BayesNet.h [hunk 5]
```diff
@@ -86,6 +87,13 @@ class BayesNet : public FactorGraph<CONDITIONAL> {
                  const KeyFormatter& keyFormatter = DefaultKeyFormatter,
                  const DotWriter& writer = DotWriter()) const;
 
+  /// @}
+  /// @name HybridValues methods
+  /// @{
+
+  double logProbability(const HybridValues& x) const;
```
Comment (varunagrawal) [inline gtsam/inference/BayesNet.h:94]: Documentation is missing.

File: python/gtsam/tests/test_DiscreteBayesNet.py [hunk 6]
```diff
@@ -11,13 +11,15 @@
 
 # pylint: disable=no-name-in-module, invalid-name
 
+import math
```
Comment (varunagrawal) [inline python/gtsam/tests/test_DiscreteBayesNet.py:14]: You don't need math. Rather use numpy.
E.g. `np.log`.

File: python/gtsam/tests/test_GaussianBayesNet.py [hunk 7]
```diff
@@ -12,13 +12,15 @@
 
 from __future__ import print_function
 
+import math
```
Comment (varunagrawal) [inline python/gtsam/tests/test_GaussianBayesNet.py:15]: Use `numpy` instead.

**Review by varunagrawal — CHANGES_REQUESTED** [review]

File: gtsam/symbolic/SymbolicConditional.h (hunk 2 above)
Comment (dellaert) [reply gtsam/symbolic/SymbolicConditional.h:22]: Had linking errors that I could not resolve for the life of me.

File: gtsam/hybrid/HybridBayesNet.cpp (hunk 3 above)
Comment (dellaert) [reply gtsam/hybrid/HybridBayesNet.cpp:296]: Sure

File: gtsam/hybrid/HybridValues.h (hunk 4 above)
Comment (dellaert) [reply gtsam/hybrid/HybridValues.h:46]: I disagree

File: gtsam/inference/BayesNet.h (hunk 5 above)
Comment (dellaert) [reply gtsam/inference/BayesNet.h:94]: Will fix

File: gtsam/hybrid/HybridBayesNet.cpp (hunk 3 above)
Comment (dellaert) [reply gtsam/hybrid/HybridBayesNet.cpp:296]: Renamed to `result`. Note that naming convention for variables in gtsam is camlCase.

**Review by varunagrawal — APPROVED** [review]

**Pull request merged**